File: src/main/kotlin/top_150/_219_contains_duplicate_ii.java
package top_150;

import java.util.HashSet;
import java.util.Set;

public class _219_contains_duplicate_ii {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        Set<Integer> set = new HashSet<>();
        for (int i = 0; i < nums.length; i++) {
            if (i > k) set.remove(nums[i - k - 1]);
            if (!set.add(nums[i])) return true;
        }
        return false;
    }

    public static void main(String[] args) {
        var s = new _219_contains_duplicate_ii();
        System.out.println(s.containsNearbyDuplicate(new int[]{1, 2, 3, 1}, 3));
//        System.out.println(s.containsNearbyDuplicate(new int[]{1, 0, 1, 1}, 1));
//        System.out.println(s.containsNearbyDuplicate(new int[]{1, 2, 3, 1, 2, 3}, 2));
//        System.out.println(s.containsNearbyDuplicate(new int[]{99, 99}, 2));
//        System.out.println(s.containsNearbyDuplicate(new int[]{1, 4, 2, 3, 1, 2}, 3));
//        System.out.println(s.containsNearbyDuplicate(new int[]{1,2,3,1}, 3));
    }
}

File: src/main/kotlin/topic/slidingwindow/easy/_219_contains_duplicate_ii.kt
package topic.slidingwindow.easy

class _219_contains_duplicate_ii {
    fun containsNearbyDuplicate(nums: IntArray, k: Int): Boolean {
        val window = HashSet<Int>()
        var l = 0
        for (r in nums.indices) {
            if (r - l > k) window.remove(nums[l++])
            if (window.contains(nums[r])) return true
            window.add(nums[r])
        }
        return false
    }
}

fun main() {
    with(_219_contains_duplicate_ii()) {
        println(containsNearbyDuplicate(intArrayOf(1,2,3,1), 3))
        println(containsNearbyDuplicate(intArrayOf(1,0,1,1), 1))
        println(containsNearbyDuplicate(intArrayOf(1,2,3,1,2,3), 2))
        println(containsNearbyDuplicate(intArrayOf(2,2), 3))
    }
}
